---
title: "biodogma"
output: rmarkdown::html_document
vignette: >
  %\VignetteIndexEntry{biodogma}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(biodogma)
```

### Description of the package `biodogma`:

-   **Function 1:** `gen_dna_seq`

-   **Function 2:** `dna_to_rna`

-   **Function 3:** `split_codon`

mRNA sequences are read in codons or groups of three ribonucleotides, each of these coding for a specific amino acid. The purpose of `split_codon` is to split a string of an RNA sequence into a vector of codons.

#### Usage

`split_codon(rna_seq, starting_codon_position)`

##### Arguments

-   `rna_seq`: a character string containing the RNA sequence.

-   `starting_codon_position`: an int variable providing the position of the initial codon in the given sequence. It's *not necessary* to provide this argument, the default value being position 1.

##### Return

-   A vector containing the codons found in the given string.

#### Example

```{r}
library("biodogma")

# Provide an RNA sequence
rna_seq <- "AUGACUAAGA"
pos <- 2

# Call the function without specifying the starting position
split_codon(rna_seq)

# Call the function specifying the starting position
split_codon(rna_seq, pos)
```

#### Step-by-step explanation

The function `split_codon` follows these steps:

1.  Obtains the length of `rna_seq` using `nchar()` and saves it in a variable (`len`).
2.  Splits `rna_seq` into codons or non-overlapping 3-character strings, using the function `susbtring()`. In the function `susbtring()` the arguments given are the `rna_seq` string and sequences of the initial and ending positions of each codon. These sequences are defined using `seq()`, which calls both `len` and `starting_codon_position`.

-   **Function 4:** `gen_aa_seq`

-   **Function 5:** `aa_count`

Amino acid sequences contain valuable information about protein structure. The `aa_count` function helps visualize the composition of these sequences, being useful to analyze certain protein characteristics like the enrichment in certain amino acid.

#### Installation

Make sure to have the `stringr` and `ggplot2` packages installed. If not, you can install them using:

```{r}
install.packages("stringr")
install.packages("ggplot2")
```

#### Usage

`aa_count(seq)`

##### Arguments

-   `seq`: a character **string** containing the amino acid sequence.

##### Return

-   A `ggplot` bar plot showing the absolute frequencies for each amino acid present in the given sequence.

#### Example

```{r}
library("biodogma")

# Provide an amino acid sequence
seq1 <- "AKLPYNILSDFHVNSTRLGE"

# Call the function 
aa_plot <- aa_count(seq1)
```

This example produces a barplot saved in the `aa_plot` object.

#### Step-by-step explanation

The function follows these steps:

1.  Obtain a vector with unique amino acids in the sequence.
2.  Counts amino acids using `str_count` from the `stringr` package.
3.  Creates data frame with the frequencies of each amino acid.
4.  Plot: using `ggplot2` a bar plot is generated showing the absolute frequencies.

### Discussion

-   Use cases for the package and what other functions could be included.

-   Task 3 - Group discussion

**Dependency:**

Although it is usually better to avoid dependencies, in this package we have used functions from `stringr` and `ggplot2` packages, due to their improved functionality and great usage. The avoidance of dependencies simplifies the complexity of the maintenance of packages as new changes in other packages may affect the functionality of another. Furthermore, it increases the usage as it does not need additional packages besides the one you are installing. However, this is not always possible as packages like those integrating the tidyverse implement complex function in an efficient manner.

In this package we have used `@importFrom package function` tags in the function description because it reduces the repetition of the `package::function` form, however the latter does not add it to the namespace and reduces the dependencies as it only loads the used functions and not the whole package (although this can also be achieved by using `min_version = TRUE` in the other form).
